File: packages/go/registry.go
package aikit

import (
	"context"
	"sort"
	"strings"
	"sync"
	"time"
)

type registryEntry struct {
	expires   time.Time
	fetchedAt time.Time
	data      []ModelMetadata
}

type learnedEntry struct {
	expires time.Time
	reason  string
}

type registryKey struct {
	Provider    Provider
	Fingerprint string
	AccountID   string
	Region      string
	Environment string
	TenantID    string
	UserID      string
}

type learnedKey struct {
	registryKey
	ModelID string
}

type modelRegistry struct {
	adapters    map[Provider]ProviderAdapter
	factory     AdapterFactory
	ttl         time.Duration
	learnedTTL  time.Duration
	cache       map[registryKey]registryEntry
	learned     map[learnedKey]learnedEntry
	mu          sync.RWMutex
}

func newModelRegistry(adapters map[Provider]ProviderAdapter, ttl time.Duration, factory AdapterFactory) *modelRegistry {
	return &modelRegistry{
		adapters:   adapters,
		factory:    factory,
		ttl:       ttl,
		learnedTTL: 20 * time.Minute,
		cache:      make(map[registryKey]registryEntry),
		learned:    make(map[learnedKey]learnedEntry),
	}
}

func (r *modelRegistry) List(ctx context.Context, opts *ListModelsOptions) ([]ModelMetadata, error) {
	entries, err := r.entriesForProviders(ctx, opts)
	if err != nil {
		return nil, err
	}
	results := make([]ModelMetadata, 0)
	for _, entry := range entries {
		results = append(results, entry.data...)
	}
	sort.Slice(results, func(i, j int) bool {
		if results[i].Provider == results[j].Provider {
			return results[i].DisplayName < results[j].DisplayName
		}
		return results[i].Provider < results[j].Provider
	})
	return results, nil
}

func (r *modelRegistry) ListRecords(ctx context.Context, opts *ListModelsOptions) ([]ModelRecord, error) {
	entries, err := r.entriesForProviders(ctx, opts)
	if err != nil {
		return nil, err
	}
	var entitlement *EntitlementContext
	if opts != nil {
		entitlement = opts.Entitlement
	}
	results := make([]ModelRecord, 0)
	for provider, entry := range entries {
		for _, model := range entry.data {
			results = append(results, r.modelRecordFromMetadata(model, provider, entry.fetchedAt, entitlement))
		}
	}
	sort.Slice(results, func(i, j int) bool {
		if results[i].Provider == results[j].Provider {
			return results[i].DisplayName < results[j].DisplayName
		}
		return results[i].Provider < results[j].Provider
	})
	return results, nil
}

func (r *modelRegistry) LearnModelUnavailable(entitlement *EntitlementContext, provider Provider, modelID string, err error) {
	reason, ok := learnReason(err)
	if !ok {
		return
	}
	key := r.learnedKey(provider, entitlement, modelID)
	r.mu.Lock()
	r.learned[key] = learnedEntry{
		expires: time.Now().Add(r.learnedTTL),
		reason:  reason,
	}
	r.mu.Unlock()
}

func (r *modelRegistry) entriesForProviders(ctx context.Context, opts *ListModelsOptions) (map[Provider]registryEntry, error) {
	providers := r.resolveProviders(opts)
	results := make(map[Provider]registryEntry, len(providers))
	for _, provider := range providers {
		entry, err := r.modelsForProvider(ctx, provider, opts)
		if err != nil {
			return nil, err
		}
		results[provider] = entry
	}
	return results, nil
}

func (r *modelRegistry) resolveProviders(opts *ListModelsOptions) []Provider {
	if opts != nil && len(opts.Providers) > 0 {
		return opts.Providers
	}
	if opts != nil && opts.Entitlement != nil && opts.Entitlement.Provider != "" {
		return []Provider{opts.Entitlement.Provider}
	}
	providers := make([]Provider, 0, len(r.adapters))
	for provider := range r.adapters {
		providers = append(providers, provider)
	}
	return providers
}

func (r *modelRegistry) modelsForProvider(ctx context.Context, provider Provider, opts *ListModelsOptions) (registryEntry, error) {
	refresh := opts != nil && opts.Refresh
	var entitlement *EntitlementContext
	if opts != nil {
		entitlement = opts.Entitlement
	}
	key := r.registryKey(provider, entitlement)
	if !refresh {
		if cached, ok := r.cached(key); ok {
			return cached, nil
		}
	}
	entry, err := r.fetchAndCache(ctx, provider, entitlement, key)
	if err != nil {
		if cached, ok := r.cached(key); ok {
			return cached, nil
		}
		return registryEntry{}, err
	}
	return entry, nil
}

func (r *modelRegistry) cached(key registryKey) (registryEntry, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	entry, ok := r.cache[key]
	if !ok || time.Now().After(entry.expires) {
		return registryEntry{}, false
	}
	return entry, true
}

func (r *modelRegistry) fetchAndCache(ctx context.Context, provider Provider, entitlement *EntitlementContext, key registryKey) (registryEntry, error) {
	adapter, err := r.adapterFor(provider, entitlement)
	if err != nil {
		return registryEntry{}, err
	}
	models, err := adapter.ListModels(ctx)
	if err != nil {
		return registryEntry{}, err
	}
	for idx, model := range models {
		models[idx] = applyCuratedMetadata(model)
	}
	now := time.Now()
	entry := registryEntry{
		data:      models,
		expires:   now.Add(r.ttl),
		fetchedAt: now,
	}
	r.mu.Lock()
	r.cache[key] = entry
	r.mu.Unlock()
	return entry, nil
}

func (r *modelRegistry) adapterFor(provider Provider, entitlement *EntitlementContext) (ProviderAdapter, error) {
	if r.factory != nil {
		return r.factory(provider, entitlement)
	}
	if adapter, ok := r.adapters[provider]; ok {
		return adapter, nil
	}
	return nil, &KitError{
		Kind:     ErrorValidation,
		Message:  "provider not configured",
		Provider: provider,
	}
}

func (r *modelRegistry) registryKey(provider Provider, entitlement *EntitlementContext) registryKey {
	key := registryKey{
		Provider:    provider,
		Fingerprint: "default",
	}
	if entitlement == nil {
		return key
	}
	key.AccountID = strings.TrimSpace(entitlement.AccountID)
	key.Region = strings.TrimSpace(entitlement.Region)
	key.Environment = strings.TrimSpace(entitlement.Environment)
	key.TenantID = strings.TrimSpace(entitlement.TenantID)
	key.UserID = strings.TrimSpace(entitlement.UserID)
	fingerprint := strings.TrimSpace(entitlement.APIKeyFingerprint)
	if fingerprint == "" && strings.TrimSpace(entitlement.APIKey) != "" {
		fingerprint = FingerprintAPIKey(entitlement.APIKey)
	}
	if fingerprint != "" {
		key.Fingerprint = fingerprint
	}
	return key
}

func (r *modelRegistry) learnedKey(provider Provider, entitlement *EntitlementContext, modelID string) learnedKey {
	return learnedKey{
		registryKey: r.registryKey(provider, entitlement),
		ModelID:     modelID,
	}
}

func (r *modelRegistry) learnedStatus(provider Provider, entitlement *EntitlementContext, modelID string) (learnedEntry, bool) {
	key := r.learnedKey(provider, entitlement, modelID)
	r.mu.Lock()
	defer r.mu.Unlock()
	entry, ok := r.learned[key]
	if !ok {
		return learnedEntry{}, false
	}
	if time.Now().After(entry.expires) {
		delete(r.learned, key)
		return learnedEntry{}, false
	}
	return entry, true
}

func (r *modelRegistry) modelRecordFromMetadata(model ModelMetadata, provider Provider, verifiedAt time.Time, entitlement *EntitlementContext) ModelRecord {
	modalities := ModelModalities{
		Text:   model.Capabilities.Text,
		Vision: model.Capabilities.Vision,
		AudioIn: model.Capabilities.AudioIn,
		AudioOut: model.Capabilities.AudioOut,
		ImageOut: model.Capabilities.Image,
		VideoIn: model.Capabilities.VideoIn,
		VideoOut: model.Capabilities.Video,
	}
	features := ModelFeatures{
		Tools:      model.Capabilities.ToolUse,
		JSONMode:   model.Capabilities.StructuredOutput,
		JSONSchema: model.Capabilities.StructuredOutput,
		Streaming:  true,
	}
	var limits *ModelLimits
	if model.ContextWindow > 0 {
		limits = &ModelLimits{ContextTokens: model.ContextWindow}
	}
	var pricing *ModelPricing
	if model.TokenPrices != nil {
		pricing = &ModelPricing{
			Currency:    "USD",
			InputPer1M:  model.TokenPrices.Input,
			OutputPer1M: model.TokenPrices.Output,
			Source:      "config",
		}
	}
	tags := []string{}
	if model.InPreview {
		tags = append(tags, "preview")
	}
	if model.Deprecated {
		tags = append(tags, "deprecated")
	}
	availability := ModelAvailability{
		Entitled:   true,
		Confidence: AvailabilityListed,
	}
	if !verifiedAt.IsZero() {
		availability.LastVerifiedAt = verifiedAt.UTC().Format(time.RFC3339)
	}
	if learned, ok := r.learnedStatus(provider, entitlement, model.ID); ok {
		availability.Entitled = false
		availability.Confidence = AvailabilityLearned
		availability.Reason = learned.reason
	}
	recordID := string(provider) + ":" + model.ID
	return ModelRecord{
		ID:             recordID,
		Provider:       provider,
		ProviderModelID: model.ID,
		DisplayName:    model.DisplayName,
		Modalities:     modalities,
		Features:       features,
		Limits:         limits,
		Tags:           tags,
		Pricing:        pricing,
		Availability:   availability,
	}
}

func learnReason(err error) (string, bool) {
	if err == nil {
		return "", false
	}
	if kitErr, ok := err.(*KitError); ok {
		if kitErr.Kind == ErrorProviderNotFound || kitErr.Kind == ErrorValidation {
			return kitErr.Message, true
		}
		if kitErr.UpstreamStatus == 400 || kitErr.UpstreamStatus == 404 || kitErr.UpstreamStatus == 403 {
			return kitErr.Message, true
		}
	}
	return "", false
}

File: packages/python/src/ai_kit/registry.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Dict, Iterable, List, Optional, Tuple

from .entitlements import fingerprint_api_key
from .errors import ErrorKind, KitErrorPayload, AiKitError
from .pricing import apply_curated_metadata
from .types import (
    EntitlementContext,
    ModelAvailability,
    ModelFeatures,
    ModelLimits,
    ModelMetadata,
    ModelModalities,
    ModelPricing,
    ModelRecord,
    Provider,
)


@dataclass
class CacheEntry:
    data: List[ModelMetadata]
    fetched_at: float
    expires_at: float


class ModelRegistry:
    def __init__(
        self,
        adapters: Dict[Provider, object],
        ttl_seconds: int = 1800,
        learned_ttl_seconds: int = 1200,
        adapter_factory=None,
    ) -> None:
        self._adapters = adapters
        self._adapter_factory = adapter_factory
        self._ttl_seconds = ttl_seconds
        self._learned_ttl_seconds = learned_ttl_seconds
        self._cache: Dict[str, CacheEntry] = {}
        self._learned: Dict[str, Tuple[float, str]] = {}

    def list_models(
        self,
        providers: Optional[List[Provider]] = None,
        refresh: bool = False,
        entitlement: Optional[EntitlementContext] = None,
    ) -> List[ModelMetadata]:
        entries = self._entries_for_providers(providers, refresh, entitlement)
        models = []
        for entry in entries.values():
            models.extend(entry.data)
        return sorted(models, key=lambda m: (m.provider, m.displayName))

    def list_model_records(
        self,
        providers: Optional[List[Provider]] = None,
        refresh: bool = False,
        entitlement: Optional[EntitlementContext] = None,
    ) -> List[ModelRecord]:
        entries = self._entries_for_providers(providers, refresh, entitlement)
        records: List[ModelRecord] = []
        for provider, entry in entries.items():
            for model in entry.data:
                records.append(
                    self._to_record(model, provider, entry.fetched_at, entitlement)
                )
        return sorted(records, key=lambda r: (r.provider, r.displayName or ""))

    def learn_model_unavailable(
        self,
        entitlement: Optional[EntitlementContext],
        provider: Provider,
        model_id: str,
        err: Exception,
    ) -> None:
        reason = _learn_reason(err)
        if not reason:
            return
        key = self._learned_key(provider, entitlement, model_id)
        self._learned[key] = (
            _now_timestamp() + self._learned_ttl_seconds,
            reason,
        )

    def _entries_for_providers(
        self,
        providers: Optional[List[Provider]],
        refresh: bool,
        entitlement: Optional[EntitlementContext],
    ) -> Dict[Provider, CacheEntry]:
        resolved = self._resolve_providers(providers, entitlement)
        if not resolved:
            raise AiKitError(
                KitErrorPayload(
                    kind=ErrorKind.VALIDATION,
                    message="No providers configured",
                )
            )
        results: Dict[Provider, CacheEntry] = {}
        for provider in resolved:
            results[provider] = self._for_provider(provider, refresh, entitlement)
        return results

    def _resolve_providers(
        self,
        providers: Optional[List[Provider]],
        entitlement: Optional[EntitlementContext],
    ) -> List[Provider]:
        if providers:
            return providers
        if entitlement and entitlement.provider:
            return [entitlement.provider]
        return list(self._adapters.keys())

    def _for_provider(
        self,
        provider: Provider,
        refresh: bool,
        entitlement: Optional[EntitlementContext],
    ) -> CacheEntry:
        key = self._cache_key(provider, entitlement)
        if not refresh:
            cached = self._cache.get(key)
            if cached and cached.expires_at > _now_timestamp():
                return cached
        try:
            entry = self._fetch_and_cache(provider, entitlement, key)
            return entry
        except Exception as err:
            cached = self._cache.get(key)
            if cached and cached.expires_at > _now_timestamp():
                return cached
            raise err

    def _fetch_and_cache(
        self,
        provider: Provider,
        entitlement: Optional[EntitlementContext],
        key: str,
    ) -> CacheEntry:
        adapter = self._adapter_for(provider, entitlement)
        if not adapter:
            raise AiKitError(
                KitErrorPayload(
                    kind=ErrorKind.VALIDATION,
                    message=f"Provider {provider} is not configured",
                )
            )
        models = adapter.list_models()
        models = [apply_curated_metadata(model) for model in models]
        now = _now_timestamp()
        entry = CacheEntry(
            data=models,
            fetched_at=now,
            expires_at=now + self._ttl_seconds,
        )
        self._cache[key] = entry
        return entry

    def _adapter_for(self, provider: Provider, entitlement: Optional[EntitlementContext]):
        if self._adapter_factory:
            return self._adapter_factory(provider, entitlement)
        return self._adapters.get(provider)

    def _cache_key(self, provider: Provider, entitlement: Optional[EntitlementContext]) -> str:
        fingerprint = "default"
        if entitlement:
            fingerprint = entitlement.apiKeyFingerprint or fingerprint_api_key(entitlement.apiKey) or "default"
        return "|".join(
            [
                provider,
                fingerprint,
                (entitlement.accountId if entitlement else "") or "",
                (entitlement.region if entitlement else "") or "",
                (entitlement.environment if entitlement else "") or "",
                (entitlement.tenantId if entitlement else "") or "",
                (entitlement.userId if entitlement else "") or "",
            ]
        )

    def _learned_key(
        self,
        provider: Provider,
        entitlement: Optional[EntitlementContext],
        model_id: str,
    ) -> str:
        return f"{self._cache_key(provider, entitlement)}|{model_id}"

    def _learned_status(
        self,
        provider: Provider,
        entitlement: Optional[EntitlementContext],
        model_id: str,
    ) -> Optional[str]:
        key = self._learned_key(provider, entitlement, model_id)
        entry = self._learned.get(key)
        if not entry:
            return None
        expires_at, reason = entry
        if expires_at < _now_timestamp():
            self._learned.pop(key, None)
            return None
        return reason

    def _to_record(
        self,
        model: ModelMetadata,
        provider: Provider,
        fetched_at: float,
        entitlement: Optional[EntitlementContext],
    ) -> ModelRecord:
        modalities = ModelModalities(
            text=model.capabilities.text,
            vision=model.capabilities.vision,
            audioIn=getattr(model.capabilities, "audio_in", None),
            audioOut=getattr(model.capabilities, "audio_out", None),
            imageOut=getattr(model.capabilities, "image", None),
        )
        features = ModelFeatures(
            tools=model.capabilities.tool_use,
            jsonMode=model.capabilities.structured_output,
            jsonSchema=model.capabilities.structured_output,
            streaming=True,
        )
        limits = (
            ModelLimits(contextTokens=model.contextWindow)
            if model.contextWindow
            else None
        )
        pricing = (
            ModelPricing(
                currency="USD",
                inputPer1M=model.tokenPrices.input if model.tokenPrices else None,
                outputPer1M=model.tokenPrices.output if model.tokenPrices else None,
                source="config",
            )
            if model.tokenPrices
            else None
        )
        tags = []
        if model.inPreview:
            tags.append("preview")
        if model.deprecated:
            tags.append("deprecated")
        availability = ModelAvailability(
            entitled=True,
            confidence="listed",
            lastVerifiedAt=_to_iso(fetched_at),
        )
        learned_reason = self._learned_status(provider, entitlement, model.id)
        if learned_reason:
            availability.entitled = False
            availability.confidence = "learned"
            availability.reason = learned_reason
        return ModelRecord(
            id=f"{provider}:{model.id}",
            provider=provider,
            providerModelId=model.id,
            displayName=model.displayName,
            modalities=modalities,
            features=features,
            limits=limits,
            tags=tags or None,
            pricing=pricing,
            availability=availability,
        )


def _now_timestamp() -> float:
    return datetime.now(tz=timezone.utc).timestamp()


def _to_iso(timestamp: float) -> str:
    return datetime.fromtimestamp(timestamp, tz=timezone.utc).isoformat()


def _learn_reason(err: Exception) -> Optional[str]:
    if isinstance(err, AiKitError):
        if err.kind in (ErrorKind.PROVIDER_NOT_FOUND, ErrorKind.VALIDATION):
            return str(err)
        if err.upstreamStatus in (400, 403, 404):
            return str(err)
    return None
